Update the quickstart guide for using Elastic Agent in a Kubernetes cluster (#2433)
elastic/docs-content: solutions/observability/get-started/quickstart-monitor-kubernetes-cluster-with-elastic-agent.md

* [**Kibana**](/deploy-manage/users-roles/cluster-or-deployment-auth/kibana-privileges.md): `{ spaces: ['*'], feature: { fleet: ['all'], fleetv2: ['all'] } }`
:::

* A running Kubernetes cluster with internet access
* [Kubectl](https://kubernetes.io/docs/reference/kubectl/)
::::

::::{tab-item} Serverless
:sync: serverless

* An {{obs-serverless}} project. To learn more, refer to [Create an Observability project](/solutions/observability/get-started.md).
* A user with the **Admin** role or higher (required to onboard system logs and metrics). To learn more, refer to [Assign user roles and privileges](/deploy-manage/users-roles/cloud-organization/user-roles.md#general-assign-user-roles).
* A running Kubernetes cluster with internet access
* [Helm](https://helm.sh/docs/)
::::

:::::

## Limitations [monitor-k8s-with-agent-limitations]

The installation command provided by the UI during the quickstart cannot be used as-is to install {{agent}} in an air-gapped environment. For an air-gapped environment with a self-managed {{stack}} deployment or orchestrator such as [{{eck}}](/deploy-manage/deploy/cloud-on-k8s.md), refer to the following resources:

> - [Air-gapped installation of Elastic Agent](/deploy-manage/deploy/self-managed/air-gapped-install.md#air-gapped-elastic-agent)

Author: You’re right, thanks!

- [Install Elastic Agent on Kubernetes using Helm](/reference/fleet/install-on-kubernetes-using-helm.md)
- [Deploy Elastic Agent in standalone mode with ECK](/deploy-manage/deploy/cloud-on-k8s/standalone-elastic-agent.md)

> - [Running Elastic Agents in an air-gapped environment](/reference/fleet/air-gapped.md)

Reviewer: ```suggestion
- [Run Elastic Agent in an air-gapped environment](/reference/fleet/air-gapped.md)
```

:sync: stack

1. In {{kib}}, go to the **Observability** UI and click **Add Data**.
2. In the **What do you want to monitor?** section, select **Kubernetes**, and then select **Elastic Agent: Logs & Metrics**.

> 3. To install the {{agent}} on your host, copy and run the install command.

Author: That’s a good point. Sometimes the definitive article is necessary when describing a specific situation with a single/specific agent, but you’re right that here it should be just Elastic Agent.

>     By running this command, you use `kubectl kustomize` to download a manifest file, inject deployment-specific data such as the API key generated by {{kib}} for the acting user, and create the Kubernetes resources for {{agent}}.

Reviewer 2: @vishaangelova @eedugon please keep me honest here, but AFAIK the helm chart supports both Elastic Stack and Serverless,  actually it can't tell the difference. So why in one we propose the kustomize approach, which we are thinking of deprecating?

Reviewer 2: oh I see thanks @eedugon

Author: Yes, the quickstart on the Kibana UI shows the `kubectl kustomize` command, but that will be changed to the `helm` one in version 9.2. I’ve opened an [issue](https://github.com/elastic/docs-content/issues/2456) to update this section when we get to 9.2 (but we’ll still have to have the text for `kubectl` because the quickstart command provided by the UI in versions 9.0 and 9.1 will still have `kubectl`)...

4. Go back to the **Kubernetes: Logs & Metrics** page in {{kib}}.

    There might be a slight delay before data is ingested. When ready, you will see the message **We are monitoring your cluster**.

5. Click **Explore Kubernetes cluster** to navigate to dashboards and explore your data.

:::

::::{tab-item} Serverless
:sync: serverless

1. Go to your [{{obs-serverless}}](/solutions/observability/get-started.md) project, then go to **Add Data**.
2. In the **What do you want to monitor?** section, select **Kubernetes**, and then select **Elastic Agent: Logs & Metrics**.
3. To install the {{agent}} on your host, copy and run the install command.

>     You will use the Helm package manager to install and configure an instance of the {{agent}} Helm chart with additional deployment-specific data such as the API key generated by {{kib}} for the acting user.

Reviewer: ```suggestion
    Use the Helm package manager to install and configure an instance of the {{agent}} Helm chart with additional deployment-specific data such as the API key generated by {{kib}} for the acting user.
```

>     The Helm chart also includes a default installation of [`kube-state-metrics` (KSM)](https://github.com/kubernetes/kube-state-metrics) as it is required by the Kubernetes integration to collect cluster-level metrics.

Reviewer: ```suggestion
    The Helm chart also includes a default installation of [`kube-state-metrics` (KSM)](https://github.com/kubernetes/kube-state-metrics), which is required by the Kubernetes integration to collect cluster-level metrics.
```

    If you encounter an error during the installation, refer to [Troubleshooting](#monitor-k8s-with-agent-troubleshooting).

    :::{dropdown} Details about the install command
    The install command provided by the UI may be similar to:

    ```sh
    helm repo add elastic https://helm.elastic.co/ && helm install elastic-agent elastic/elastic-agent --version 9.1.0 -n kube-system
    ```

    Where:
    
    - `elastic` is the name of the local Helm configuration of the Elastic Helm repository (`https://helm.elastic.co/`).
    - `elastic-agent` is the name of the specific installation of the Helm chart, known as **release name**.
    - `elastic/elastic-agent` is the Helm chart to be installed.

>     - `9.1.0` is the version of the {{agent}} Helm chart to be installed.

Author: Currently subs are not supported in inline code, but I used the sub without the code formatting here.

    - `kube-system` is the namespace where {{agent}} is installed.

>     The command will also include `--set` configurations specific to your deployment and user. Refer to [Install standalone Elastic Agent on Kubernetes using Helm](/reference/fleet/example-kubernetes-standalone-agent-helm.md#agent-standalone-helm-example-install) for a more detailed explanation of the configuration options used.

Reviewer 3: Consider something like this:

```
helm repo add elastic https://helm.elastic.co/ && helm install elastic-agent elastic/elastic-agent --version 9.1.1 -n kube-system --set outputs.default.url=https:<elasticsearch-url>:443 --set kubernetes.onboardingID=<internal-id> --set kubernetes.enabled=true --set outputs.default.type=ESPlainAuthAPI --set outputs.default.api_key=$(echo "<api-key>" | base64 -d)
```

:::::{tab-set}
:group: stack-serverless

::::{tab-item} Elastic Stack
:sync: stack

To uninstall the {{agent}} and the Kubernetes resources installed with `kubectl`:

1. Copy the `kubectl` quickstart command for installing {{agent}} described in the [Collect your data](#_collect_your_data_2) section.
2. Replace `| kubectl apply -f-` with `| kubectl delete -f-`, then run the command.
::::

::::{tab-item} Serverless
:sync: serverless

To uninstall the {{agent}} and the Kubernetes resources installed with Helm, run: 

  ```sh
  helm uninstall <release-name> -n <namespace> <1>
  ```
  1. Substitute `<release-name>` with the release name and `<namespace>` with the namespace used in the quickstart command described in the [Collect your data](#_collect_your_data_2) section.

:::{dropdown} Expand to view an example
If the {{agent}} Helm chart is installed with the `elastic-agent` release name in the `kube-system` namespace, use the following command to uninstall it:

```
helm uninstall elastic-agent -n kube-system
```
:::

::::

:::::

## Troubleshooting [monitor-k8s-with-agent-troubleshooting]

### `kube-state-metrics` is already installed

> If you're using `helm` to install {{agent}} in your Kubernetes cluster, you may encounter an error if `kube-state-metrics` is already installed in the same namespace where the {{agent}} is to be installed. In this case, add the option `--set kube-state-metrics.enabled=false` to the install command provided by the UI to skip the installation of `kube-state-metrics`.

Reviewer: ```suggestion
If you're using `helm` to install {{agent}} in your Kubernetes cluster, you may encounter an error if `kube-state-metrics` is already installed in the same namespace where {{agent}} is to be installed. In this case, add the option `--set kube-state-metrics.enabled=false` to the install command provided by the UI to skip the installation of `kube-state-metrics`.
```
